### ISSUE TITLE
 Are the start and end parameter types of torch.linspace operators wrong?

### ISSUE THREAD

**Issue by wenhaopeter** [pytorch/pytorch#72788]
### 📚 The doc issue

The start and end parameters of the torch.linspace operator support torch.int32 and torch.int64 input, and can correctly output the corresponding type of tensor, but it is in https://pytorch.org/docs/master/generated/torch.linspace.html#torch.linspace, the start and end parameters only support float. Is the documentation wrong?

>>> import torch
>>>a=torch.linspace(23453245323,23453245333,steps=5,dtype=torch.int64)
>>> a
tensor([23453245323, 23453245325, 23453245328, 23453245330, 23453245333])

**Comment by samdow** [comment]
I believe that what's happening is that start/end inputs must be promotable to float, so though the start and ending can be ints, those will be promoted to float. As an additional note, the start and ends don't have to be ints for the output to be an int, that's done at the end: ex:
```
>>> torch.linspace(5.5, 10.5, steps=4, dtype=torch.int64)
tensor([ 5,  6,  8, 10])
```

**Comment by gchanan** [comment]
should the type in the documentation just be `Number`, as it is in arange?

**Comment by mruberry** [comment]
@gchanan is correct. The type annotation is incorrect, and linspace supports integer, floating point, and complex start and stop values. The type should just be Number. We would take a correction for this. It should also correct the logspace documentation, which likely has a similar issue.

**Comment by wenhaopeter** [comment]
The comment of the document is indeed wrong, but I have another problem. If I input a relatively large int value, such as more than 2**53, the data type is torch.int64, but the output tensor result is not what I want. for example:

>>>import torch  
>>>torch.linspace(1729382256910270500,1729382256910270600,steps=10,dtype=torch.int64)  
tensor([1729382256910270464, 1729382256910270464, 1729382256910270464,  
        1729382256910270464, 1729382256910270464, 1729382256910270464,  
        1729382256910270464, 1729382256910270464, 1729382256910270720,  
        1729382256910270720])  

By reading the source code, I found that step is a floating point type, and start is an int type. The addition results in conversion to a floating point type, which will cause the start to be incorrect. Does this need to be fixed?

**Comment by mruberry** [comment]
I'm not sure what bug you're seeing in your latest comment, @wenhaopeter? Could you elaborate? It looks like the output tensor has the correct `torch.int64` datatype.

**Comment by wenhaopeter** [comment]
The output tensor dtype is correct, but resulut is wrong.The right result is [1729382256910270500, 729382256910270510, 1729382256910270520,1729382256910270540,1729382256910270550,1729382256910270560,1729382256910270570,1729382256910270580,1729382256910270590,1729382256910270600],but we get wrong.

The reason is that the process of start+step is converted to float, resulting in loss of precision

**Comment by mruberry** [comment]
Aha! Thanks for clarifying @wenhaopeter. @pmeier has volunteered to investigate a fix.

**Comment by pmeier** [comment]
Just as FYI: this might take a while since I'm currently busy with other stuff. If someone else wants to take this up, feel free to do so.
